### PR TITLE
Hardcode error type in IntoVisitor

### PR DESCRIPTION
**Pull request by jsdw**
This PR hardcodes the error type, because, in a nutshell, we run into this issue in some places otherwise:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=51e845f1f73ed72b0eb9d3d29172089a

We can avoid _that_ specific issue by not having any bounds on `DecodeAsType` and instaed only bounding implementations of `DecodeAsType`, but that just means that `DecodeAsType` doesn't imply `IntoVisitor`, and so code like the following fails:

```rust
#[derive(DecodeAsType)]
#[decode_as_type(trait_bounds = "T: DecodeAsType")]
struct Foo<T> {
    val: Option<T>
}
```

This fails because for `Option<T>` to implement `DecodeAsType`, it must implement `IntoVisitor`, and it only does so if `T: IntoVisitor, T::Visitor::Error: Into<Error>`, ie we can't just use `DecodeAsType` as a trait bound.

With a hardcoded `Error` type on `IntoVisitor`, `DecodeAsType` can imply `IntoVisitor`, and the above issue goes away. I also added a `VisitorWithCrateError` adapter struct which makes any visitor whose error is convertible into `crate::Error` into a visitor returning `crate::Error`, so that it's easy to still adapt any viable visitor into something that can be used with `IntoVisitor`.

### CODE REVIEW

File: scale-decode/src/visitor/mod.rs [hunk 1]
```diff
@@ -358,6 +378,33 @@ impl Visitor for IgnoreVisitor {
...
+    fn unchecked_decode_as_type<'scale, 'info>(
+        self,
+        input: &mut &'scale [u8],
+        type_id: TypeId,
+        types: &'info scale_info::PortableRegistry,
+    ) -> DecodeAsTypeResult<Self, Result<Self::Value<'scale, 'info>, Self::Error>> {
+        let res = decode_with_visitor(input, type_id.0, types, self.0).map_err(Into::into);
+        DecodeAsTypeResult::Decoded(res)
+    }
+}
```
Comment (tadeohepperle) [inline scale-decode/src/visitor/mod.rs:407]: Nice, this will allow for all sorts of easy conversions.

**Review by tadeohepperle — APPROVED** [review]

File: scale-decode/src/visitor/mod.rs (hunk 1 above)
Comment (jsdw) [reply scale-decode/src/visitor/mod.rs:407]: Yeah; it basically replaces the "auto conversion" we did before but has the nice effect of simplifying the code anyway and making `DecodeAsType` work better, so I'm quite happy with this solution in the end :)

**Review by niklasad1 — APPROVED** [review]
LGTM

**Review by lexnv — APPROVED** [review]
Nice one!

**Pull request merged**